### PR TITLE
Requeue operations from blocks when re-org occurs

### LINKED ISSUE #2001: Re-queue operations from blocks when a re-org occurs
### Description
To better maintain the pending pools of operations to include in proposed blocks:

 1. When blocks are imported, only remove the operations they include from pools if the block is on our canonical chain
 2. When a re-org occurs:
    a. re-add operations from the old chain to the pending pools
    b. remove operations from the new chain from pending pools

Operations refer to items included in the block body - attestations, slashings and voluntary exits.

### PR DESCRIPTION
**Pull request by cemozerr**
## PR Description
This PR deals with the management of operations in re-org situations.

Specifically, when blocks are imported, removes the operations from operation pools only if the block is on our canonical chain, and when a re-org occurs:
-  re-adds operations from the old chain to the operation pools
-  removes operations from the new chain from operation pools

## Fixed Issue(s)

Fixes #2001

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationsReOrgManager.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+                                                  LOG.error(
+                                                      "Failed to process re-queued attestation from block: {} due to: {}",
+                                                      root,
+                                                      err));
+                                    });
+                          },
+                          () ->
+                              LOG.warn(
+                                  "Failed to re-queue operations for now non-canonical block: {}",
+                                  root)))
```
Comment (ajsutton) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationsReOrgManager.java]: I don't think this is worth a warn level - if the block was finalized during the processing it isn't a big deal if we don't recover the operations from it.

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/blockimport/BlockImporter.java [hunk 2]
```diff
@@ -82,7 +83,13 @@ public BlockImporter(
 
               final Optional<BlockProcessingRecord> record = result.getBlockProcessingRecord();
               eventBus.post(new ImportedBlockEvent(block));
-              notifyBlockOperationSubscribers(block);
+
+              // Notify operation pools to remove operations only
+              // if the block is on our canonical chain
+              if (((SuccessfulBlockImportResult) result).isBlockOnCanonicalChain()) {
```
Comment (ajsutton) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/blockimport/BlockImporter.java]: We shouldn't be casting to `SuccesfulBlockImportResult` - it's always just a `BlockImportResult` and for failures `isBlockOnCanonicalChain` would always return false (block isn't on the canonical chain if it failed to import).

File: storage/src/main/java/tech/pegasys/teku/storage/client/RecentChainData.java [hunk 3]
```diff
@@ -186,6 +187,12 @@ public UpdatableStore getStore() {
         .orElseGet(TreeMap::new);
   }
 
+  public NavigableMap<UInt64, Bytes32> getAncestorRootsForRoot(
+      final UInt64 startSlot, Bytes32 root) {
+    return ForkChoiceUtil.getAncestors(
+        forkChoiceStrategy.orElseThrow(), root, startSlot, UInt64.ONE, UInt64.valueOf(10000));
```
Comment (ajsutton) [inline storage/src/main/java/tech/pegasys/teku/storage/client/RecentChainData.java]: Hard coded 10000 doesn't look right here.

Comment (ajsutton) [reply storage/src/main/java/tech/pegasys/teku/storage/client/RecentChainData.java]: Also I'm suspicious that we might be including the common ancestor block itself when processing a reorg.  We should process both chains only from the block after the common ancestor (ie only process the blocks that actually changed).

File: ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/OperationsReOrgManagerTest.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+import tech.pegasys.teku.storage.client.ChainUpdater;
+import tech.pegasys.teku.storage.client.RecentChainData;
+import tech.pegasys.teku.storage.storageSystem.InMemoryStorageSystemBuilder;
+import tech.pegasys.teku.storage.storageSystem.StorageSystem;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+
+public class OperationsReOrgManagerTest {
```
Comment (ajsutton) [inline ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/OperationsReOrgManagerTest.java:47]: This looks awfully complex and I think it still has to setup quite a lot of operations yet.  I suspect getting it all set up with valid data will be a lot of code and pretty slow.  I think I'd just do a unit test focused on OperationsReOrgManager - with a mock RecentChainData it should be fairly easy to inject the blocks we want and all we really need to see is that the operations from those blocks and added or removed from all the right places.  We know they do their job fine from there.

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationsReOrgManager.java (hunk 1 above)
Comment (cemozerr) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationsReOrgManager.java]: Makes sense.

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/blockimport/BlockImporter.java (hunk 2 above)
Comment (cemozerr) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/blockimport/BlockImporter.java]: Makes sense.

File: storage/src/main/java/tech/pegasys/teku/storage/client/RecentChainData.java (hunk 3 above)
Comment (cemozerr) [reply storage/src/main/java/tech/pegasys/teku/storage/client/RecentChainData.java]: Yeah, I wasn't sure about 10000 as well. I need a number high enough to never be a problem, but less than the limits of UInt64 so that it never throws an `ArithmeticException` as well.

Comment (cemozerr) [reply storage/src/main/java/tech/pegasys/teku/storage/client/RecentChainData.java]: Added a new method to get every ancestor that does not calculate `count` instead of passing a magic variable.

Comment (cemozerr) [reply storage/src/main/java/tech/pegasys/teku/storage/client/RecentChainData.java]: Also made sure this method does not return the block root for the common ancestor.

File: ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/OperationsReOrgManagerTest.java (hunk 4 above)
Comment (cemozerr) [reply ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/OperationsReOrgManagerTest.java:47]: Makes sense. Will switch to unit tests.

File: ethereum/core/src/main/java/tech/pegasys/teku/core/ForkChoiceUtil.java [hunk 5]
```diff
@@ -119,6 +119,26 @@ public static UInt64 compute_slots_since_epoch_start(UInt64 slot) {
     return roots;
   }
 
+  /**
+   * @param forkChoiceStrategy
+   * @param root
+   * @param startSlot
+   * @return every block root from root (inclusive) to start slot (exclusive) traversing the chain
+   *     backwards
+   */
```
Comment (ajsutton) [inline ethereum/core/src/main/java/tech/pegasys/teku/core/ForkChoiceUtil.java]: If we're going to add a java doc comment we should fill it all out. :)

I also feel like this should be some variant of `getAncestors` so that it has a similar name to the existing `getAncestors` method.  Maybe `getAncestorsOnFork`?

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationsReOrgManager.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+                              LOG.debug(
+                                  "Failed to re-queue operations for now non-canonical block: {}",
+                                  root)))
+              .finish(
+                  err ->
+                      LOG.warn(
+                          "Failed to re-queue operations for now non-canonical block: {} due to future error: {}",
+                          root,
+                          err.getMessage()));
+        });
```
Comment (ajsutton) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationsReOrgManager.java]: The nesting levels here seem to have gotten a bit out of control.  Probably need to extract a method or two to reduce them.

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationsReOrgManager.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+                          () ->
+                              LOG.debug(
+                                  "Failed to re-queue operations for now non-canonical block: {}",
+                                  root)))
+              .finish(
+                  err ->
+                      LOG.warn(
+                          "Failed to re-queue operations for now non-canonical block: {} due to future error: {}",
+                          root,
+                          err.getMessage()));
```
Comment (ajsutton) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationsReOrgManager.java]: We should preserve the stack trace instead of only logging the error message.
```suggestion
                      LOG.warn(
                          "Failed to re-queue operations for now non-canonical block: {}",
                          root,
                          err));
```

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationsReOrgManager.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+                                                      reason ->
+                                                          LOG.debug(
+                                                              "Rejected re-queued attestation from block: {} due to: {}",
+                                                              root,
+                                                              reason)),
+                                              err ->
+                                                  LOG.error(
+                                                      "Failed to process re-queued attestation from block: {} due to: {}",
+                                                      root,
+                                                      err));
```
Comment (ajsutton) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationsReOrgManager.java]:
```suggestion
                                                  LOG.error(
                                                      "Failed to process re-queued attestation from block: {}",
                                                      root,
                                                      err));
```

Avoid losing the error message.

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationsReOrgManager.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    this.exitPool = exitPool;
+    this.proposerSlashingPool = proposerSlashingPool;
+    this.attesterSlashingPool = attesterSlashingPool;
+    this.attestationManager = attestationManager;
+    this.attestationPool = attestationPool;
+    this.recentChainData = recentChainData;
+  }
+
+  @Override
+  public void reorgOccurred(
```
Comment (ajsutton) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationsReOrgManager.java:63]: This method is too big overall. Probably worth splitting in two - process nonCanonical and then process canonical blocks.  And then probably a few other helper methods to make the optional/future chainings not wind up so deeply nested and hard to follow.

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/attestation/AttestationManager.java [hunk 10]
```diff
@@ -112,7 +113,10 @@ private void onBlockImported(final ImportedBlockEvent blockImportedEvent) {
                               "Failed to process pending attestation dependent on " + blockRoot,
                               err));
             });
-    block.getMessage().getBody().getAttestations().forEach(aggregatingAttestationPool::remove);
+  }
+
+  public SafeFuture<AttestationProcessingResult> onAttestation(Attestation attestation) {
+    return onAttestation(ValidateableAttestation.fromAttestation(attestation));
   }
```
Comment (ajsutton) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/attestation/AttestationManager.java]: I probably wouldn't add this method.  We want to create the `ValidatableAttestation` only at the key entry points but this makes it much simpler to get that wrong and create a new `ValidatableAttestation` without thinking about it and then we'll lose the context it tracks.

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java [hunk 11]
```diff
@@ -146,15 +147,18 @@ private void processHead(Optional<UInt64> nodeSlot) {
                       forkChoiceStrategy.onAttestation(transaction, indexedAttestation));
           return transaction
               .commit()
-              .thenRun(() -> updateForkChoiceForImportedBlock(block, forkChoiceStrategy, result))
+              .thenRun(
+                  () ->
+                      updateForkChoiceForImportedBlock(
+                          block, forkChoiceStrategy, (SuccessfulBlockImportResult) result))
```
Comment (ajsutton) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java]: Should avoid the cast to `SuccessfulBlockImportResult`.  Might need to make `setBlockOnCanonicalChain` a `markAsCanonical` method which throws `UnsupportedOperationException` for failed results.

File: ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/OperationsReOrgManagerTest.java [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+      new OperationsReOrgManager(
+          proposerSlashingOperationPool,
+          attesterSlashingOperationPool,
+          exitOperationPool,
+          attestationPool,
+          attestationManager,
+          recentChainData);
+
+  @Test
+  void shouldRequeueAndRemoveOperations() throws Exception {
```
Comment (ajsutton) [inline ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/OperationsReOrgManagerTest.java]: Probably worth having a test for when blocks are removed but not added or vice-versa (ie a reorg that only changes from an empty slot to a full one and the other way).  Should work fine but that's the most common type of reorg so important to handle it right.

File: storage/src/main/java/tech/pegasys/teku/storage/client/CombinedChainDataClient.java [hunk 13]
```diff
@@ -357,7 +357,7 @@ public UpdatableStore getStore() {
 
   public NavigableMap<UInt64, Bytes32> getAncestorRoots(
       final UInt64 startSlot, final UInt64 step, final UInt64 count) {
-    return recentChainData.getAncestorRoots(startSlot, step, count);
+    return recentChainData.getAncestorRootsForChainHead(startSlot, step, count);
```
Comment (ajsutton) [inline storage/src/main/java/tech/pegasys/teku/storage/client/CombinedChainDataClient.java]: Should rename this to match the recentChainData method it delegates to.

File: storage/src/test/java/tech/pegasys/teku/storage/client/RecentChainDataTest.java [hunk 14]
```diff
@@ -564,6 +578,47 @@ public void commit_pruneParallelExistingBlocks() throws Exception {
     testCommitPruningOfParallelBlocks(false);
   }
 
+  @Test
+  public void getEveryRootOnChainTillSlot() {
```
Comment (ajsutton) [inline storage/src/test/java/tech/pegasys/teku/storage/client/RecentChainDataTest.java]: Probably should have a test case for when the block root is unknown.

**Review by ajsutton — COMMENTED** [review]
Looking good.

File: ethereum/core/src/main/java/tech/pegasys/teku/core/ForkChoiceUtil.java (hunk 5 above)
Comment (cemozerr) [reply ethereum/core/src/main/java/tech/pegasys/teku/core/ForkChoiceUtil.java]: Makes sense. Done.

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationsReOrgManager.java (hunk 9 above)
Comment (cemozerr) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationsReOrgManager.java:63]: I tried to make it better, and its a bit better, but there is still some serious nesting. Could use suggestions.

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationsReOrgManager.java (hunk 8 above)
Comment (cemozerr) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationsReOrgManager.java]: Done.

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationsReOrgManager.java (hunk 7 above)
Comment (cemozerr) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationsReOrgManager.java]: Done.

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationsReOrgManager.java (hunk 6 above)
Comment (cemozerr) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationsReOrgManager.java]: Done.

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/attestation/AttestationManager.java (hunk 10 above)
Comment (cemozerr) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/attestation/AttestationManager.java]: Makes sense.

Comment (cemozerr) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/attestation/AttestationManager.java]: Done.

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java (hunk 11 above)
Comment (cemozerr) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java]: Makes sense. Done.

File: ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/OperationsReOrgManagerTest.java (hunk 12 above)
Comment (cemozerr) [reply ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/OperationsReOrgManagerTest.java]: I did the reOrg from an empty slot to a full one, but I'm having a hard time understanding the reOrg from full one to empty slot since attestations vote on blocks.

File: storage/src/main/java/tech/pegasys/teku/storage/client/CombinedChainDataClient.java (hunk 13 above)
Comment (cemozerr) [reply storage/src/main/java/tech/pegasys/teku/storage/client/CombinedChainDataClient.java]: Done.

File: storage/src/test/java/tech/pegasys/teku/storage/client/RecentChainDataTest.java (hunk 14 above)
Comment (cemozerr) [reply storage/src/test/java/tech/pegasys/teku/storage/client/RecentChainDataTest.java]: Done.

File: data/serializer/src/main/java/tech/pegasys/teku/api/schema/AttesterSlashing.java [hunk 15]
```diff
@@ -23,6 +26,14 @@ public AttesterSlashing(
...
   }
 
+  @JsonCreator
+  public AttesterSlashing(
+      @JsonProperty("attestation_1") final IndexedAttestation attestation_1,
+      @JsonProperty("attestation_2") final IndexedAttestation attestation_2) {
+    this.attestation_1 = attestation_1;
+    this.attestation_2 = attestation_2;
+  }
+
```
Comment (ajsutton) [inline data/serializer/src/main/java/tech/pegasys/teku/api/schema/AttesterSlashing.java:36]: I'm not sure why this change was required as part of this PR.

File: data/serializer/src/main/java/tech/pegasys/teku/api/schema/IndexedAttestation.java [hunk 16]
```diff
@@ -40,6 +42,16 @@ public IndexedAttestation(
...
 
+  @JsonCreator
+  public IndexedAttestation(
+      @JsonProperty("attesting_indices") final List<UInt64> attesting_indices,
+      @JsonProperty("data") final AttestationData data,
+      @JsonProperty("signature") final BLSSignature signature) {
+    this.attesting_indices = attesting_indices;
+    this.data = data;
+    this.signature = signature;
+  }
```
Comment (ajsutton) [inline data/serializer/src/main/java/tech/pegasys/teku/api/schema/IndexedAttestation.java:53]: Not sure what caused this to be needed either.

File: ethereum/core/src/main/java/tech/pegasys/teku/core/ForkChoiceUtil.java [hunk 17]
```diff
@@ -119,6 +119,26 @@ public static UInt64 compute_slots_since_epoch_start(UInt64 slot) {
...
+   * @param startSlot the slot (exclusive) until which we walk the chain backwards
+   * @return every block root from root (inclusive) to start slot (exclusive) traversing the chain
+   *     backwards
+   */
+  public static NavigableMap<UInt64, Bytes32> getAncestorsOnFork(
+      ForkChoiceStrategy forkChoiceStrategy, Bytes32 root, UInt64 startSlot) {
+    final NavigableMap<UInt64, Bytes32> roots = new TreeMap<>();
+    Bytes32 parentRoot = root;
+    Optional<UInt64> parentSlot = forkChoiceStrategy.blockSlot(parentRoot);
+    while (parentSlot.isPresent() && parentSlot.get().compareTo(startSlot) > 0) {
```
Comment (ajsutton) [inline ethereum/core/src/main/java/tech/pegasys/teku/core/ForkChoiceUtil.java]:
```suggestion
    while (parentSlot.isPresent() && parentSlot.get().isGreaterThan(startSlot)) {
```

File: ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/operations/AttesterSlashing.java [hunk 18]
```diff
@@ -41,6 +42,7 @@
   private final IndexedAttestation attestation_2;
 
   @Label("sos-ignore")
+  @JsonIgnore
```
Comment (ajsutton) [inline ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/operations/AttesterSlashing.java]: We definitely shouldn't have JSON serialisation annotations in our core classes.

File: ethereum/datastructures/src/testFixtures/java/tech/pegasys/teku/datastructures/util/DataStructureUtil.java [hunk 19]
```diff
@@ -126,6 +126,11 @@ public BLSSignature randomSignature() {
     return randomSSZList(classInfo, maxSize / 10, maxSize, valueGenerator);
   }
 
+  public <T> SSZList<T> randomSSZList(
+      Class<? extends T> classInfo, long maxSize, Supplier<T> valueGenerator, long numItems) {
+    return randomSSZList(classInfo, numItems, maxSize, valueGenerator);
+  }
```
Comment (ajsutton) [inline ethereum/datastructures/src/testFixtures/java/tech/pegasys/teku/datastructures/util/DataStructureUtil.java:132]: Why is this a new method just to reorder the params?

File: ethereum/datastructures/src/testFixtures/java/tech/pegasys/teku/datastructures/util/DataStructureUtil.java [hunk 20]
```diff
@@ -381,8 +386,11 @@ public BeaconBlockBody randomBeaconBlockBody() {
         randomSSZList(
             ProposerSlashing.class, Constants.MAX_PROPOSER_SLASHINGS, this::randomProposerSlashing),
         randomSSZList(
-            AttesterSlashing.class, Constants.MAX_ATTESTER_SLASHINGS, this::randomAttesterSlashing),
-        randomSSZList(Attestation.class, Constants.MAX_ATTESTATIONS, this::randomAttestation),
+            AttesterSlashing.class,
+            Constants.MAX_ATTESTER_SLASHINGS,
+            this::randomAttesterSlashing,
+            1),
+        randomSSZList(Attestation.class, Constants.MAX_ATTESTATIONS, this::randomAttestation, 1),
```
Comment (ajsutton) [inline ethereum/datastructures/src/testFixtures/java/tech/pegasys/teku/datastructures/util/DataStructureUtil.java]: I'd probably have at least 3 attestations in the blocks by default.  Good to have a slashing added in there that I'm guessing wound up with none before though.

**Review by ajsutton — APPROVED** [review]
LGTM.  Just a few seemingly unrelated JSON changes as well.  The ones to API seem unnecessary but the new annotation in operations/AttesterSlashing we definitely want to get rid of as we deliberately avoid having the JSON stuff through the internal data structures.

File: data/serializer/src/main/java/tech/pegasys/teku/api/schema/AttesterSlashing.java (hunk 15 above)
Comment (ajsutton) [reply data/serializer/src/main/java/tech/pegasys/teku/api/schema/AttesterSlashing.java:36]: Ok I think I can see this one being useful for the submit block API which previously was a bug.  Still not sure what's deserializing an `IndexedAttestation` since they should only be in states and we shouldn't parse any of those from JSON (I think).

Comment (cemozerr) [reply data/serializer/src/main/java/tech/pegasys/teku/api/schema/AttesterSlashing.java:36]: We need to parse `IndexedAttestation` when we're deserializing JSON of `AttesterSlashing`, since it contains two `IndexedAttestation`s.

I needed to do deal with all this JSON stuff as I made sure now our `DataStructureUtil.randomBeaconBlock(int n)` method always returns a block with 1 `AttesterSlashing`, instead of returning an empty `AttesterSlashing` `SSZList` as it used to do. This made all our tests more robust IMO.

Then it turned out that we could not deserialize `BeaconBlock` JSON objects which included `AttesterSlashings`. I saw this as the `PostBlockTest.shouldReturnAcceptedIfBlockFailsValidation` failed after creating and trying to deserialize a randomBeaconBlock, which now included an AttesterSlashing.

File: ethereum/core/src/main/java/tech/pegasys/teku/core/ForkChoiceUtil.java (hunk 17 above)
Comment (cemozerr) [reply ethereum/core/src/main/java/tech/pegasys/teku/core/ForkChoiceUtil.java]: Done.

File: ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/operations/AttesterSlashing.java (hunk 18 above)
Comment (cemozerr) [reply ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/operations/AttesterSlashing.java]: Artifact of intense debugging. Removed.

File: ethereum/datastructures/src/testFixtures/java/tech/pegasys/teku/datastructures/util/DataStructureUtil.java (hunk 19 above)
Comment (cemozerr) [reply ethereum/datastructures/src/testFixtures/java/tech/pegasys/teku/datastructures/util/DataStructureUtil.java:132]: Its to have a configurable number of items in the `SSZList`. In the past, we either had the `randomFullSSZList` option which filled the list to its `maxSize`, or the default `randomSSZList`, which would fill each list `maxSize/10` number of items. 

In attester slashing, that would mean 0 items.

File: ethereum/datastructures/src/testFixtures/java/tech/pegasys/teku/datastructures/util/DataStructureUtil.java (hunk 20 above)
Comment (cemozerr) [reply ethereum/datastructures/src/testFixtures/java/tech/pegasys/teku/datastructures/util/DataStructureUtil.java]: Yep exactly the case.

File: data/serializer/src/main/java/tech/pegasys/teku/api/schema/AttesterSlashing.java (hunk 15 above)
Comment (ajsutton) [reply data/serializer/src/main/java/tech/pegasys/teku/api/schema/AttesterSlashing.java:36]: oh right of course. Yep that makes sense.
